### PR TITLE
Hive: Support column projection on Hive Profiles

### PR DESCRIPTION
**Pull request by frankgh**
Currently, Hive and HiveRC profiles do not support column
projection. Add support for column projection for these profiles at the
PXF level. PXF can determine the following scenarios:

1. The number of columns of the file is higher than the number of
   columns defined at the Greenplum table definition.
2. The number of columns of the file is less than the number of columns
   defined at the Greenplum table definition.
3. A mixed number of columns is present in the files we are accessing.

For this to work, the column names on the Greenplum table definition
must match the column names on the Hive table definition, as the
projection is based on column names rather than column positions.

HiveText will not support column projection.

### CODE REVIEW

File: automation/src/test/java/org/greenplum/pxf/automation/features/hive/HiveRcTest.java [hunk 1]
```diff
@@ -244,7 +331,7 @@ public void severalPartitionsDefaultSerde() throws Exception {
      *
      * @throws Exception if test fails to run
      */
-    @Test(groups = { "hive", "features", "gpdb", "security" })
+//    @Test(groups = { "hive", "features", "gpdb", "security" })
```
[the remark this answers is not in the document]
Comment (frankgh) [reply automation/src/test/java/org/greenplum/pxf/automation/features/hive/HiveRcTest.java:334]: Removing this test as we now support querying a subset of Hive columns in Greenplum

File: automation/src/test/java/org/greenplum/pxf/automation/features/hive/HiveTest.java [hunk 2]
```diff
@@ -517,7 +577,8 @@ public void incorrectProfile() throws Exception {
      *
      * @throws Exception if test fails to run
      */
-    @Test(groups = {"hive", "features", "gpdb", "security"})
+    //@Test(groups = {"hive", "features", "gpdb", "security"})
+    @Ignore("we now support column count mismatch for Hive")
```
[the remark this answers is not in the document]
Comment (frankgh) [reply automation/src/test/java/org/greenplum/pxf/automation/features/hive/HiveTest.java:581]: Removing this test as this is now supported

File: server/pxf-hive/src/main/java/org/greenplum/pxf/plugins/hive/HiveResolver.java [hunk 3]
```diff
@@ -463,36 +467,62 @@ private void traverseTuple(Object obj, ObjectInspector objInspector,
...
-                    fields.get(i).getFieldObjectInspector(), complexRecord,
-                    toFlatten);
-            if (toFlatten) {
-                addOneFieldToRecord(structRecord, DataType.TEXT,
-                        HdfsUtilities.toString(complexRecord, mapkeyDelim));
+        OneField partitionField;
+
+        if (toFlatten) {
+            for (int i = 0; i < structFields.size(); i++) {
+                complexRecord.add(new OneField(DataType.TEXT.getOID(), String.format("\"%s\"", fields.get(i).getFieldName())));
```
[the remark this answers is not in the document]
Comment (frankgh) [reply server/pxf-hive/src/main/java/org/greenplum/pxf/plugins/hive/HiveResolver.java:479]: Yes, flattening does not happen for tuples https://github.com/greenplum-db/pxf/blob/4e363c4ef2bd2a488895586ca6cab8736069f162/server/pxf-hive/src/main/java/org/greenplum/pxf/plugins/hive/HiveResolver.java#L131 but rather on nested structs https://github.com/greenplum-db/pxf/blob/4e363c4ef2bd2a488895586ca6cab8736069f162/server/pxf-hive/src/main/java/org/greenplum/pxf/plugins/hive/HiveResolver.java#L413 . The flattening happens on columns of type collection for example here: https://github.com/greenplum-db/pxf/blob/4e363c4ef2bd2a488895586ca6cab8736069f162/automation/tincrepo/main/pxf/features/hive/hive_collection_types/expected/query01.ans#L6 Take a look at column t2, t3, t4, and t5, where t2 is a ARRAY<STRING>, t3 a MAP<STRING, FLOAT>, t4 a STRUCT<street:STRING, city:STRING, state:STRING, zip:INT> (this column specifically will call the traverseStruct for it's value with a flatten = true value), t5 a UNIONTYPE<STRING, INT, ARRAY<INT>, ARRAY<STRING>

File: server/pxf-hive/src/main/java/org/greenplum/pxf/plugins/hive/HiveResolver.java [hunk 4]
```diff
@@ -463,36 +467,62 @@ private void traverseTuple(Object obj, ObjectInspector objInspector,
...
+            }
         }
+
         return toFlatten ? structRecord : complexRecord;
     }
 
+    private OneField getPartitionField(String columnName) {
+        OneField oneField = partitionColumnNames.get(columnName);
+        if (oneField == null) {
+            oneField = partitionColumnNames.get(columnName.toLowerCase());
```
[the remark this answers is not in the document]
Comment (frankgh) [reply server/pxf-hive/src/main/java/org/greenplum/pxf/plugins/hive/HiveResolver.java:521]: This was mostly to support a very edge case. But I am reading that the Hive metastore stores the table schema in all lowercase. So it does not apply. I will fix it

File: server/pxf-hive/src/main/java/org/greenplum/pxf/plugins/hive/HiveResolver.java [hunk 5]
```diff
@@ -463,36 +467,62 @@ private void traverseTuple(Object obj, ObjectInspector objInspector,
...
+                    IntStream.range(0, fields.size())
+                            .boxed()
+                            .collect(Collectors.toMap(i -> fields.get(i).getFieldName(), i -> i));
+
+            List<ColumnDescriptor> tupleDescription = context.getTupleDescription();
+            for (int j = 0; j < tupleDescription.size(); j++) {
+                ColumnDescriptor columnDescriptor = tupleDescription.get(j);
+                Integer i = defaultIfNull(columnNameToStructIndexMap.get(columnDescriptor.columnName()),
+                        columnNameToStructIndexMap.get(columnDescriptor.columnName().toLowerCase()));
+                Integer structIndex = hiveIndexes.get(j);
```
[the remark this answers is not in the document]
Comment (frankgh) [reply server/pxf-hive/src/main/java/org/greenplum/pxf/plugins/hive/HiveResolver.java:495]: good question, I will add a note here, in case we need it

**Comment by ururu-fy** [comment]
@frankgh 
Hello. Does it support HiveORC profile?

**Comment by frankgh** [comment]
@ururu-fy it supports projection for `HiveORC` also

**Pull request merged**